### PR TITLE
migrate stretch histogram to use plot plugin-subcomponent

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request migrates the stretch histogram in plot options to make use of the plugin plot subcomponent introduced in #2254.

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/configs/default/plugins/plot_options/plot_options.py [hunk 1]
```diff
@@ -191,8 +188,8 @@ class PlotOptions(PluginTemplateMixin):
     stretch_vmax_sync = Dict().tag(sync=True)
 
     stretch_hist_zoom_limits = Bool().tag(sync=True)
-    stretch_hist_nbins = IntHandleEmpty().tag(sync=True)
-    stretch_histogram = Any().tag(sync=True, **widget_serialization)
+    stretch_hist_nbins = IntHandleEmpty(25).tag(sync=True)
```
Comment (pllim) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:191]: Where did 25 come from?

File: jdaviz/configs/default/plugins/plot_options/plot_options.py [hunk 2]
```diff
@@ -374,6 +371,16 @@ def state_attr_for_line_visible(state):
...
+        self.stretch_histogram = Plot(self)
+        self.stretch_histogram.add_bins('histogram', color='gray')
+        self.stretch_histogram.add_line('vmin', x=[0, 0], y=[0, 1], ynorm=True, color='#c75d2c')
+        self.stretch_histogram.add_line('vmax', x=[0, 0], y=[0, 1], ynorm=True, color='#c75d2c')
+        self.stretch_histogram.figure.axes[0].label = 'pixel value'
+        self.stretch_histogram.figure.axes[0].num_ticks = 3
+        self.stretch_histogram.figure.axes[0].tick_format = '0.1e'
+        self.stretch_histogram.figure.axes[1].label = 'density'
+        self.stretch_histogram.figure.axes[1].num_ticks = 2
+        self.stretch_histogram_widget = 'IPY_MODEL_'+self.stretch_histogram.model_id
```
Comment (pllim) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:383]:
```suggestion
        self.stretch_histogram_widget = f'IPY_MODEL_{self.stretch_histogram.model_id}'
```

File: jdaviz/configs/default/plugins/plot_options/plot_options.py [hunk 3]
```diff
@@ -579,96 +585,34 @@ def _update_stretch_histogram(self, msg={}):
...
 
-        self._check_if_v_stretch_changed()
-
-    @observe('stretch_vmin_value', 'stretch_vmax_value')
-    def _check_if_v_stretch_changed(self, msg=None):
-        self._remove_histogram_marks()
-        self._add_histogram_marks()
+    @observe('stretch_vmin_value')
+    def _stretch_vmin_changed(self, msg=None):
+        self.stretch_histogram.marks['vmin'].x = [self.stretch_vmin_value] * 2
```
Comment (pllim) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:600]: No need for multiplication overhead when it is just 2?
```suggestion
        self.stretch_histogram.marks['vmin'].x = [self.stretch_vmin_value, self.stretch_vmin_value]
```

File: jdaviz/configs/default/plugins/plot_options/plot_options.py [hunk 4]
```diff
@@ -579,96 +585,34 @@ def _update_stretch_histogram(self, msg={}):
...
-        self.stretch_histogram.marks = self.stretch_histogram.marks + v_stretch_lines
-
-    def _remove_histogram_marks(self):
-        if self.stretch_histogram is None:
-            return
-        self.stretch_histogram.marks = [mark for mark in self.stretch_histogram.marks
-                                        if not isinstance(mark, HistogramMark)]
+    @observe('stretch_vmax_value')
+    def _stretch_vmax_changed(self, msg=None):
+        self.stretch_histogram.marks['vmax'].x = [self.stretch_vmax_value] * 2
```
Comment (pllim) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:604]:
```suggestion
        self.stretch_histogram.marks['vmax'].x = [self.stretch_vmax_value, self.stretch_vmax_value]
```

File: jdaviz/configs/default/plugins/plot_options/plot_options.py [hunk 5]
```diff
@@ -579,96 +585,34 @@ def _update_stretch_histogram(self, msg={}):
...
 
-    @observe("stretch_hist_nbins")
-    def histogram_nbins_changed(self, msg):
+    @observe('stretch_hist_nbins')
```
Comment (pllim) [inline jdaviz/configs/default/plugins/plot_options/plot_options.py:606]: There is actually a movement to prefer double quotes but let's not open that can of worms here. But then again, no reason for this change?
```suggestion
    @observe("stretch_hist_nbins")
```

**Review by pllim — COMMENTED** [review]
General nitpicks first.

File: jdaviz/configs/default/plugins/plot_options/plot_options.py (hunk 1 above)
Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:191]: The previous implementation used 25 bins by default in https://github.com/spacetelescope/jdaviz/pull/2314/files#diff-522ee60e0cfb6e1041a540ed4c2a07cd8efc622ca39f31bd55af884d0cc4d96dL590

File: jdaviz/configs/default/plugins/plot_options/plot_options.py (hunk 5 above)
Comment (kecnry) [reply jdaviz/configs/default/plugins/plot_options/plot_options.py:606]: I was just trying to be self-consistent between the three observes in a row - but no sense in adding to the diff (maybe I'll switch the new one instead).

**Review by pllim — APPROVED** [review]
I saw the bug where histogram bars are not showing until I resize the plugin, but I think you said it is unrelated? So I am approving. Thanks!

**Review by javerbukh — APPROVED** [review]
Looks great, nice work!

**Comment by kecnry** [comment]
(Let's hold off on merge for a bit now that we're deferring the work that builds off this and I'll try to investigate the bug where the plot when opening the tray manually doesn't always update until a resize event).

File: jdaviz/core/template_mixin.py [hunk 6]
```diff
@@ -2681,7 +2684,8 @@ def add_scatter(self, label, x=[], y=[], xnorm=False, ynorm=False, **kwargs):
                               colors=kwargs.pop('color', kwargs.pop('colors', 'gray')),
                               **kwargs)
 
-    def add_bins(self, label, sample=[1], bins=1, density=True, **kwargs):
+    def add_bins(self, label, sample=[0], bins=2, density=True, **kwargs):
+        # NOTE: initializing with bins=1 breaks the figure until a resize event
```
Comment (kecnry) [inline jdaviz/core/template_mixin.py]: @pllim @javerbukh - this was the actual fix for the tray-resize issue.  I kept a few other small changes from attempts to fix this that I think will still be more efficient, but I don't expect them to be very controversial 🤞

Comment (pllim) [reply jdaviz/core/template_mixin.py]: My approval still stands. I don't see how this is going to break anything new. Thanks!

Comment (javerbukh) [reply jdaviz/core/template_mixin.py]: My approval stills stands as well!

**Pull request merged**